### PR TITLE
test(compiler): add tests for the terminal-logger module

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This adds tests for the terminal-logger module and also does a little
bit of code cleanup (some light refactoring, JSDoc, etc).

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Previously the `terminal-logger` implementation didn't have any tests around it 😱 

## What is the new behavior?

This PR adds some tests and does some light refactoring. You can see the history a bit by stepping through the commits, basically what I ended up doing was:

1. Make the changes necessary to get a `Logger` (as returned by `createTerminalLogger`) stood up in a test environment
2. add some tests for the existing functionality
3. do a few small refactors that seemed like obvious wins in terms of code clarity

There is a _lot_ more than could be done here, including looking at just tearing this whole thing out and using a library. But short of that, there is code in the module to do syntax highlighting, some weird indirection for the ANSI color escape code functions, and some more stuff that could potentially be DRY'ed up a little bit. Hopefully this moves the needle on test coverage for us in a helpful way though!

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

There should be no changes in the behavior of the code in question.

## Testing

I added tests!

### CODE REVIEW

File: src/compiler/sys/logger/test/terminal-logger.spec.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+      const { logger, writeLogsMock } = setup();
+      logger.setLogFilePath!('testfile.txt');
+      logger.info('test message');
+      logger.writeLogs!(false);
+      const expectedLogfile = [
+        '09:32:32.00  0010.0MB  I  test message',
+        '09:32:32.00  0010.0MB  F  --------------------------------------',
+      ].join('\n');
+      expect(writeLogsMock).toBeCalledWith('testfile.txt', expectedLogfile, false);
+    });
```
Comment (rwaskiewicz) [inline src/compiler/sys/logger/test/terminal-logger.spec.ts:109]: Can we add a small test that we don't write to a file if the log file path isn't set?

```suggestion
    });
    
    it("doesn't write to any logs if the log path isn't set", function () {
      const { logger, writeLogsMock } = setup();
      logger.setLogFilePath!('');
      logger.info('test message');
      logger.writeLogs!(false);
      expect(writeLogsMock).not.toHaveBeenCalled();
    });
```

File: src/compiler/sys/logger/test/terminal-logger.spec.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        timespan.finish('finish the timespan');
+
+        expect(logMock).toHaveBeenNthCalledWith(1, `${dim('[32:32.0]')}  start the timespan ${dim('...')}`);
+        expect(logMock).toHaveBeenNthCalledWith(
+          2,
+          `${dim('[32:32.0]')}  finish the timespan ${dim('in less than 1 ms')}`
+        );
+      });
+
+      it('writes timespanes to the log file, if configured', () => {
```
Comment (rwaskiewicz) [inline src/compiler/sys/logger/test/terminal-logger.spec.ts:192]:
```suggestion
      it('writes timespans to the log file, if configured', () => {
```

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz comments should be addressed!

File: src/cli/task-generate.ts [hunk 3]
```diff
@@ -70,7 +70,6 @@ export const taskGenerate = async (coreCompiler: CoreCompiler, config: Config):
     return config.sys.exit(1);
   }
 
-  // TODO(STENCIL-424): Investigate moving these console.log calls to config.logger.info
```
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/cli/task-generate.ts:73]: That makes sense, can we put a comment above this block of `console.log` to explain this/why we're not using the logger instance on `config` in that case?

[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/cli/task-generate.ts:73]: sure, that makes sense

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz updated!

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**